collections/nemo_nlp/nemo_nlp/data/datasets/bert_pretraining.py

```diff
@@ -158,66 +155,127 @@ def get_document(filepath, line):
-            b_document = get_document(b_filename, b_line)
-            b_start_idx = random.randrange(len(b_document) - b_length)
+            b_filename = a_filename
+            b_line_idx = a_line_idx + 1
+            b_document = get_document(
+                a_filename, self.sentence_indices[a_filename][b_line_idx])
+            b_document, b_line_idx = match_target_seq_length(
+                b_document, target_seq_length_b, b_filename, b_line_idx,
+                self.sentence_indices)
+
+        def truncate_seq_pair(a_document, b_document, max_num_tokens):
+            # Truncates a pair of sequences to a maximum sequence length
+            while True:
+                total_length = len(a_document) + len(b_document)
+                if total_length <= max_num_tokens:
+                    break
+
+                trunc_document = a_document if len(
+                    a_document) > len(b_document) else b_document
+                assert len(trunc_document) >= 1
+
+                if random.random() < 0.5:
+                    del trunc_document[0]
+                else:
+                    trunc_document.pop()
 
-        # Process retrieved documents for use in training
-        a_ids = a_document[a_start_idx:a_start_idx + a_length]
-        b_ids = b_document[b_start_idx:b_start_idx + b_length]
+        truncate_seq_pair(a_document, b_document, max_num_tokens)
 
-        output_ids = [self.tokenizer.special_tokens["[CLS]"]]
-        output_ids.extend(a_ids)
-        output_ids.append(self.tokenizer.special_tokens["[SEP]"])
-        output_ids.extend(b_ids)
-        output_ids.append(self.tokenizer.special_tokens["[SEP]"])
+        output_ids = [self.tokenizer.special_tokens["[CLS]"]] + a_document + \
+                     [self.tokenizer.special_tokens["[SEP]"]] + b_document + \
+                     [self.tokenizer.special_tokens["[SEP]"]]
 
         input_ids, output_mask = self.mask_ids(output_ids)
 
         output_mask = np.array(output_mask, dtype=np.float32)
-        input_mask = np.ones(self.max_seq_length, dtype=np.float32)
+        input_mask = np.zeros(self.max_seq_length, dtype=np.float32)
+        input_mask[:len(input_ids)] = 1
 
         input_type_ids = np.zeros(self.max_seq_length, dtype=np.int)
-        input_type_ids[a_length + 2:seq_length + 3] = 1
+        input_type_ids[len(a_document) + 2:len(output_ids) + 1] = 1
+
+        while len(input_ids) < self.max_seq_length:
+            input_ids.append(0)
+            output_ids.append(0)
+            output_mask = np.append(output_mask, [0])
 
-        return (np.array(input_ids),
-                input_type_ids,
-                input_mask,
-                np.array(output_ids),
-                output_mask,
-                label)
+        return np.array(input_ids), input_type_ids, np.array(
+            input_mask, dtype=np.float32), np.array(output_ids), np.array(
+            output_mask, dtype=np.float32), label
```

Comment:
suggestion: use () to wrap around the return values so that each can be in its own line. just a style choice :)

Reply:
good idea :) I'll comment the action item as a TODO for now...